[Nidhish Krishnan] iP

Nid21cs opened the pull request:
# Auntie Maggie
> "The purpose of our life is to be happy." - Dalai Lama

What can Ms Maggie do for you?
-  **Everything** 
-  She is  all you need 
-  Manage your tasks
-  ***Remember*** all your saved tasks

How to use it?
1. Download the jar file [here](https://github.com/Nid21cs/ip/releases)
1. Just run it from there :+1: !

Features:
- [x] Able to accept date inputs to keep track
- [ ] Notifications(coming soon)
- [ ] GUI (coming soon)


Have a look at the code below for some proof of how good maggie is 
```ruby
boolean bestChatBot() {
    return true;
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Nid21cs committed:
Made sure all class files are not pushed to github
Added v0 requirements

Nid21cs committed:
no message

Nid21cs committed:
Revert "no message"

This reverts commit e21044b2857cc2f0440a6f397f5fb23a8f81159e.

Nid21cs committed:
Hello added Scanner to receive input

Nid21cs committed:
Added a tasks class to keep track of given tasks and add in related methods.
Can now list out all tasks and will show when tasks are added

Nid21cs committed:
Added unmarking and marking of tasks functionality
Added single task class for each individual task

Nid21cs committed:
Used arraylist instead of using an array to put in objects of type SingleTask
Added subclasses of SingleTask to represent 3 types of Tasks.

Nid21cs committed:
Added auto testing of output specifically A-TextUiTesting

Nid21cs committed:
Added DukeException class to handle inputs that are not valid for the scannner

Nid21cs committed:
Added a delete functionality to delete tasks and also made sure that the inputs of delete are viable and throw an exception if not

Nid21cs committed:
Added a Parser class to handle user inputs, a storage class to save and load saved tasks when bot is turned off.
also made SingleTasks abstract as well

Nid21cs committed:
changed to ensure that all import statements were specific

Nid21cs committed:
Branch level 8 added a date and time feature to Deadline so the date and time is stored

Nid21cs committed:
Added a UI class to handle user input and call for it when required
Added a TaskList class to handle all methods involving the ArrayList

Nid21cs committed:
Added all files under the package class

Nid21cs committed:
Added javadoc comments to all public methods
Made sure to follow coding standard as well as
Added a find feature to find tasks that contain a certain keyword

Nid21cs committed:
Merge pull request #1 from Nid21cs/add-gradle-support

Add gradle support

Nid21cs committed:
Added gradle and is able to run

Nid21cs committed:
added junit testing

Nid21cs committed:
Added jar file to gitignore

winson8222 commented on src/main/java/Duke/Ui.java:
```diff
@@ -0,0 +1,47 @@
+package Duke;
```
Package name should be of lower case

winson8222 commented on src/test/java/Duke/DukeTest.java:
```diff
@@ -0,0 +1,17 @@
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class DukeTest {
```
Perhaps more unit tests can be made

winson8222 commented on src/main/java/Duke/TaskList.java:
```diff
@@ -0,0 +1,137 @@
+ * It provides methods to manage and manipulate the list of tasks.
+ */
+public class TaskList {
+    ArrayList<SingleTask> taskList;
```
Plural form should be used to represent a collection of objects

winson8222 commented on src/main/java/Duke/SingleTask.java:
```diff
@@ -0,0 +1,23 @@
+ * The SingleTask class represents a single task for the Duke program.
+ * It is an abstract class that provides a common interface for different types of tasks.
+ */
+public abstract class SingleTask {
```
All methods are named adhering to the coding standard.

winson8222 commented on src/main/java/Duke/Deadline.java:
```diff
@@ -0,0 +1,101 @@
+package Duke;
+
```
Consistent import statement sequence.

Nid21cs committed:
Added GUI
changed all System.print statements to return a string to output to GUI as well

MadLamprey commented on gradle/wrapper/gradle-wrapper.jar:
Maybe you could ignore this file from tracking since it is a binary file?

MadLamprey commented on src/main/java/Duke/Duke.java:
Could change file name to match the name of your bot?

MadLamprey commented on src/main/java/Duke/Parser.java:
```diff
@@ -0,0 +1,81 @@
+     * @throws DukeException if there is an error parsing the user input
+     */
+    public String parse(String s, TaskList taskList) throws DukeException {
+        if (s.matches(".*\\bdelete\\b.*")) {
```
Good use of Regex!

MadLamprey commented on src/test/java/Duke/DukeTest.java:
Could maybe write test cases for individual methods?

MadLamprey reviewed (commented):
All in all, looks good. Attention could be paid to testing.

kiatkat commented on data/duke.txt:
```diff
@@ -0,0 +1,5 @@
+T|1|balls
+T|0|sell book
+D|0|return book |2/12/2019 1800
+T|0|balls
+T|0|isha
```
Perhaps include this in the `.gitignore`, so that your stored data doesn't get included in the version control?

kiatkat commented on src/main/java/Duke/DialogBox.java:
```diff
@@ -0,0 +1,61 @@
+package Duke;
```
Shouldn't this package name be in lower case, by the coding standards? I notice this in a couple other files too.

kiatkat commented on src/main/java/Duke/Deadline.java:
```diff
@@ -0,0 +1,101 @@
+ */
+public class Deadline extends SingleTask {
+    String formattedTime;
+    LocalDateTime DueBy;
```
Should this be in PascalCase?

kiatkat commented on src/main/java/Duke/ToDo.java:
```diff
@@ -0,0 +1,76 @@
+    public String mark() {
+        this.isDone = true;
+        return "Ok boy i mark for you already \n" +
+                "[" +this.getStatusIcon() +"] " + this.description;
```
Just a small point, but I believe this is missing a whitespace between the `+` and the `this...`. This recurs in a few other places too.

kiatkat reviewed (commented):
Looks good overall! Just some minor coding style points.

Nid21cs committed:
Exceptions are only checking for bad input etcetera but not for problems in code itself assertions are not used.

Without assertions, potential bugs and issues related to task operations may go unnoticed during development and testing.

Added assertions to the three task classes as well as TaskList

Assertions are a useful way to validate assumptions about the program state. They can help catch bugs early in the development process.

Nid21cs committed:
Merge pull request #2 from Nid21cs/branch-A-Assertions

Exceptions are only checking for bad input etcetera but not for probl…

Nid21cs committed:
The code does not adhere to coding style

Coding style is important for code readability and maintainability

Corrected coding style mistakes in various files

Nid21cs committed:
Merge pull request #3 from Nid21cs/branch-A-CodeQuality

The code does not adhere to coding style

Nid21cs committed:
Added a CommandHelp class to handle all help inputs.

Returns instruction of what each command does as well as lists all available commands.

Example usage of the command is also provided

Also changed package names to adhere to coding standards.

Nid21cs committed:
Added a user guide and fixed some bugs in code

Nid21cs committed:
Updated User guide removed template

Nid21cs committed:
Added more comprehensive junit testing to test parser class.

Nid21cs committed:
Updated build.gradle to work on different OS

Nid21cs committed:
Add Testing and reduced LOC in parser.parse method

Nid21cs committed:
Updated UG

Nid21cs committed:
Added more tests and shortened methods
